enhancements/hypershift/priority-class/priority-class.md

```diff
@@ -0,0 +1,266 @@
+```
+
+With the availablity of the above priority classes, HO can be enhanced to use and override priority classes being on set on the underlying control plane workloads/pods.
+
+### Workflow Description
+TBD
+### API Extensions
+HC and HCP CRDs -- new field
+
+### Implementation Details/Notes/Constraints
+This effort will be implemented over two phases:
+- Phase 1 - Override Priority Classes
+- Phase 2 - Dynamic Adjustment of Priority Classes
+
+### Phase 1 - Override Priority Classes
+Hypershift currently creates and uses a static set of default priority classes. The first phase of this effort will largely focus on Hypershift's ability to override these priority classes for underlying workloads.
+- The overridden priority classes must be specified in the newly created API fields for the HC. 
+- The HC will have three new API fields for the three various priority classes used by hypershift currently--ControlPlanePriorityClass, APICriticalPriorityClass, and EtcdPriorityClass.
+- Override criteria:
+  - Adding a `ControlPlanePriorityClass` in the HC spec will propogate to/override any components with the `hypershift-control-plane` priority class.
+  - Adding a `APICriticalPriorityClass` in the HC spec will propogate to/override any components with the `hypershift-api-critical` priority class.
+  - Adding a `EtcdPriorityClass` in the HC spec will propogate to/override any components with the `hypershift-etcd` priority class.
+  - The `hypershift-operator`, `system-cluster-critical`, and `system-node-critical` priorty classes will not be overriden by this enhancement. 
+- The `hostedcluster_controller` will reconcile the specified override priority classes from the HC and propogate them to the HCP spec.
+- The CPO will reconcile the new override priority classes from the HCP spec, and propogate the them to the various components determined by the override criteria.
+  - The `DeploymentConfig` struct will be used as it is currently to propogate changes in priority classes.
+
+### maxSurge
+At the moment, we are mitigating a Kubernetes rollout bug that can take down the service by setting `maxSurge` to `0`. Due to the bug, we experience an outage if a pod is restarted for any reason during a rollout, and since overriding priority classes initiates a rollout, it could potentially lead to an outage. 
+
+As a temporary workaround, we've adjusted `maxSurge` values for all HA control plane deployments to `0`.
+PR: https://github.com/openshift/hypershift/issues/1568
+
+We'll need to re-evaluate if this needs to be reverted if/when the Kubernetes bug is resolved. 
+Kubernetes issue: https://github.com/openshift/hypershift/issues/1568 
+
+
+
+### Phase 2 - Dynamic Adjustment of Priority Classes (WIP)
+Phase 2 of this enhancement will focus on bolstering Hypershift's capability by introducing dynamic adjustment of priority classes based on woe workload's size. This will directly be tied to the HC's total node count.
```

Comment:
probably punt all of phase 2 to a follow-up enhancement, so you don't need to hash it out before starting in on the phase 1 stuff?

Reply:
I think we'd prefer them to stay within the same enhancement since the two phases are quite interdependent.

Reply:
We ended up employing this suggestion and splitting the phases into their own documents-thanks!